src/main/java/ca/on/oicr/pineryreports/reports/impl/PreciseReport.java

```diff
@@ -0,0 +1,607 @@
+    }
+  };
+
+  @SafeVarargs
+  private final List<List<String>> getSingleTable(List<SampleDto> possibleSamples, Map<String, SampleDto> allSamplesById,
+      List<SampleLabel> labelList, Predicate<SampleDto>... filters) {
+    List<Predicate<SampleDto>> inputFilters = Arrays.asList(filters);
+    List<SampleDto> filtered = filter(possibleSamples, inputFilters);
+    if (filtered.isEmpty()) {
+      // nothing to report on here, so skip it
+      return Collections.emptyList();
+    }
+
+    return SITE_ORDER.stream()
+        .map(site -> {
+          List<Predicate<SampleDto>> allFilters = new ArrayList<>();
+          allFilters.add(Site.predicate(site));
+          allFilters.addAll(inputFilters);
+
+          List<SampleDto> sams = filter(possibleSamples, allFilters);
+          List<String> row = new ArrayList<>();
+          row.add(site.getKey());
+          row.addAll(getCounts(sams, allSamplesById, labelList));
+          return row;
+        })
+        .collect(Collectors.toList());
+  }
+
+  private List<String> getCounts(List<SampleDto> samples, Map<String, SampleDto> samplesById, List<SampleLabel> labels) {
+    return labels.stream()
+        .map(label -> {
+          Map<String, Long> labelled = getSampleAndCaseCounts(samples, Arrays.asList(label.predicate()), samplesById);
+          return Arrays.asList(labelled.get(NUM_SAMPLES), labelled.get(NUM_CASES));
+        })
+        .flatMap(Collection::stream)
+        .map(num -> num == null ? "0" : String.valueOf(num))
+        .collect(Collectors.toList());
+  }
+
+  private Map<String, Long> getSampleAndCaseCounts(List<SampleDto> samples, Collection<Predicate<SampleDto>> predicates,
+      Map<String, SampleDto> allSamples) {
+    List<SampleDto> filteredSamples = filter(samples, predicates);
+    List<SampleDto> cases = filteredSamples.stream()
+        .map(sam -> getParent(sam, SAMPLE_CATEGORY_IDENTITY, allSamples))
+        .distinct().collect(Collectors.toList());
+
+    Map<String, Long> counts = new TreeMap<>();
+    counts.put(NUM_SAMPLES, Long.valueOf(filteredSamples.size()));
+    counts.put(NUM_CASES, Long.valueOf(cases.size()));
+    return counts;
+  }
+
+  private List<String> labelListHeader(String header, List<SampleLabel> labelList) {
+    // needs to have one column for the time point, then two columns for each sampleLabel
+    // which are for the "Samples" and "Cases" columns in the {#samplesCasesHeader} method
+    List<String> list = new ArrayList<>();
+    list.add(header);
+    for (SampleLabel label : labelList) {
+      list.add(label.getKey());
+      list.add("");
+    }
+    return list;
+  }
+
+  private List<String> sampleCasesHeader(List<SampleLabel> labelList) {
+    // first column is "Site", then a pair of "Samples" + "Cases" columns for each sampleLabel in the list
+    List<String> sampleCasesList = new ArrayList<>();
+    sampleCasesList.add("Site");
+    for (int i = 0; i < labelList.size(); i++) {
+      sampleCasesList.add("# Samples");
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Use already-defined constant 'NUM_SAMPLES' instead of duplicating its value here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1192)